crates/goose-cli/src/session.rs

```diff
@@ -252,15 +252,27 @@ We've removed the conversation up to the most recent user message
                 .push(Message::assistant().with_text(prompt_response));
             self.prompt.render(raw_message(prompt_response));
         } else {
-            // Default behavior for non-tool interruptions, remove the last user message
+            // An interruption occurred outside of a tool request-response.
             if let Some(last_msg) = self.messages.last() {
                 if last_msg.role == Role::User {
-                    self.messages.pop();
+                    match last_msg.content.first() {
+                        Some(MessageContent::ToolResponse(_)) => {
+                            // Interruption occurred after a tool had completed but not assistant reply
+                            let prompt_response = "We interrupted the existing calls to tools. How would you like to proceed?";
+                            self.messages
+                                .push(Message::assistant().with_text(prompt_response));
+                            self.prompt.render(raw_message(prompt_response));
+                        }
+                        Some(_) => {
+                            // A real users message
+                            self.messages.pop();
+                            let prompt_response = "We interrupted before the model replied and removed the last message.";
+                            self.prompt.render(raw_message(prompt_response));
+                        }
+                        None => panic!("No content in last message"),
```

Comment:
will this quit the whole agent or a child process?

Reply:
everything I imagine, but should never have a User message with no content, should be unreachable.